loguru.cpp

```diff
@@ -1100,18 +1105,20 @@ namespace loguru
 			#elif defined(__FreeBSD__)
 				long thread_id;
 				(void)thr_self(&thread_id);
+			#elif defined(__OpenBSD__)
+				pid_t thread_id = getthrid();
 			#elif LOGURU_PTHREADS
-				uint64_t thread_id = pthread_self();
+				uint64_t thread_id = reinterpret_cast<uint64_t>(pthread_self());
```

Comment:
I think you probably want `static_cast` here? If `pthread_self()` returns a 32bit, this feels like an out-of-bounds read (though maybe it just acts like a C cast…then `static_cast` is just more direct).

Reply:
Can we detect if it is a pointer and, if so, use `reinterpret_cast`? If it is an integral type, I think `static_cast` is what is wanted here.

Reply:
Is your concern that `reinterpret_cast` is a code smell to be avoided? (I agree generally)

`pthread_t` is of course supposed to be opaque so it also feels smelly to try and introspect its true nature.  It's a pointer on macOS & OpenBSD, and apparently an integer on linux. Probably it could even be a struct, in which case even `reinterpret_cast` would fail I suppose.

But I don't see any case where `reinterpret_cast` would give a different/worse result than `static_cast`.

Note also that this smelly code is in a final fallback case of a fallback case. First, there has to be nothing like `pthread_getname_np`, then if `buffer[0] == 0`, all the other fallback cases (but one) need to be unavailable too. This line should basically never run.

Ideally, this should all be replaced with STL threads, would be way more portable.

Reply:
It looks like `reinterpret_cast` does not support casting to larger integer types:

```
cast.cpp: In instantiation of ‘void print_cast_reinterpret(const char*, const char*) [with From = unsigned char; To = short unsigned int]’:
cast.cpp:105:5:   required from here                                                                       
cast.cpp:92:12: error: invalid cast from type ‘unsigned char’ to type ‘short unsigned int’                                                                                                                             
   92 |     To t = reinterpret_cast<To>(f);
      |            ^~~~~~~~~~~~~~~~~~~~~~~ 
```

So if `pthread_t` is an integral type that is not at least 64bits in size, it does not compile.

Reply:
Meh, ok, so I've just reverted that case to how it was before, but with an explicit cast at least. If there is some platform where `pthread_` fails that cast, then hopefully the compiler will catch it.

See https://github.com/emilk/loguru/pull/245/commits/6572aa15f4ebdd56b22ee28405ce4508a9a57820